### PR TITLE
Update moleculec-c2

### PR DESCRIPTION
**Pull request by XuJiandong**
This PR is part of this one:
https://github.com/nervosnetwork/ckb-production-scripts/pull/8

### CODE REVIEW

**Comment by jjyr** [comment]
https://github.com/jjyr/sparse-merkle-tree/pull/18

We can introduce the test fixtures.

File: ckb_smt.h [hunk 1]
```diff
@@ -236,43 +247,96 @@ int smt_calculate_root(uint8_t *buffer, const smt_state_t *pairs,
...
+        // push value
+        if (_smt_get_bit(key_a, height_a)) {
+          _smt_merge(height_a, parent_key, value_b, value_a, value_a);
         } else {
-          _smt_merge(value_a, value_b, value_a);
+          _smt_merge(height_a, parent_key, value_a, value_b, value_a);
         }
-        /* Top-of-stack key is already updated to parent_key_a */
+        // push key
+        _smt_parent_path(key_a, height_a);
```
Comment (TheWaWaR) [inline ckb_smt.h]: `memcpy(key_a, parent_key, SMT_KEY_BYTES)`

File: ckb_smt.h [hunk 2]
```diff
@@ -236,43 +247,96 @@ int smt_calculate_root(uint8_t *buffer, const smt_state_t *pairs,
...
+        }
+        uint8_t parent_key[SMT_KEY_BYTES];
+        memcpy(parent_key, key, SMT_KEY_BYTES);
+        uint16_t height_u16 = base_height;
+        for (uint16_t idx = 0; idx < zero_count; idx++) {
+          height_u16 = base_height + idx;
+          if (height_u16 > 255) {
+            return ERROR_INVALID_PROOF;
+          }
+          memcpy(parent_key, key, SMT_KEY_BYTES);
```
Comment (TheWaWaR) [inline ckb_smt.h]: No need to copy the `key` again, just use old `parent_key`, since height_u16 is linear increasing.

Comment (XuJiandong) [reply ckb_smt.h]: It's tricky and not necessary to do it. It's not very easy to maintain.

File: ckb_smt.h [hunk 3]
```diff
@@ -158,16 +160,19 @@ int _smt_zero_value(const uint8_t *value) {
...
 /* Notice that output might collide with one of lhs, or rhs */
-void _smt_merge(const uint8_t *lhs, const uint8_t *rhs, uint8_t *output) {
-  if (_smt_zero_value(lhs)) {
-    memcpy(output, rhs, 32);
-  } else if (_smt_zero_value(rhs)) {
-    memcpy(output, lhs, 32);
+void _smt_merge(uint8_t height, const uint8_t *node_key, const uint8_t *lhs,
+                const uint8_t *rhs, uint8_t *output) {
+  if (_smt_zero_value(lhs) && _smt_zero_value(rhs)) {
+    _smt_zero_value(output);
```
Comment (TheWaWaR) [inline ckb_smt.h]: `memcpy(output, lhs, 32)`

File: ckb_smt.h (hunk 1 above)
Comment (XuJiandong) [reply ckb_smt.h]: fixed.

File: ckb_smt.h (hunk 3 above)
Comment (XuJiandong) [reply ckb_smt.h:166]: It's same.

Comment (XuJiandong) [reply ckb_smt.h]: fixed. set it to zero.

File: ckb_smt.h (hunk 2 above)
Comment (XuJiandong) [reply ckb_smt.h]: fixed. Add  more comments

**Pull request merged**